# Úkol č. 2 - regrese

* Termíny jsou uvedeny na [courses.fit.cvut.cz](https://courses.fit.cvut.cz/BI-ML1/homeworks/index.html).
* Pokud odevzdáte úkol po prvním termínu ale před nejzazším termínem, budete penalizování -12 body, pozdější odevzdání je bez bodu.
* V rámci tohoto úkolu se musíte vypořádat s regresní úlohou, s příznaky různých typů a s chybějícími hodnotami.
* Před tím, než na nich postavíte predikční model, je třeba je nějakým způsobem převést do číselné reprezentace.
    
> **Úkoly jsou zadány tak, aby Vám daly prostor pro invenci. Vymyslet _jak přesně_ budete úkol řešit, je důležitou součástí zadání a originalita či nápaditost bude také hodnocena!**

Využívejte buňky typu `Markdown` k vysvětlování Vašeho postupu. Za nepřehlednost budeme strhávat body.

## Zdroj dat

Budeme se zabývat predikcí délky dožití v různých zemích a letech.
K dispozici máte trénovací data v souboru `data.csv` a data na vyhodnocení v souboru `evaluation.csv`.

#### Seznam příznaků:

* Year - Rok
* Status - Status rozvinuté nebo rozvojové země
* Life expectancy - Délka dožití v letech - **cílová proměnná, kterou budete predikovat**
* Adult Mortality - Úmrtnost dospělých bez ohledu na pohlaví (pravděpodobnost, že osoby, které dosáhly věku 15 let, zemřou před dosažením věku 60 let (uvedeno na 1 000 osob)).
* infant deaths - počet zemřelých kojenců na 1000 obyvatel
* Alcohol - Alkohol, zaznamenaná spotřeba na obyvatele (15+) (v litrech čistého alkoholu)
* percentage expenditure - Výdaje na zdravotnictví v procentech hrubého domácího produktu na obyvatele (%)
* Hepatitis B - pokrytí očkováním proti hepatitidě B (HepB) u dětí ve věku 1 roku (%)
* Measles - Spalničky - počet hlášených případů na 1000 obyvatel
* BMI - průměrný index tělesné hmotnosti celé populace
* under-five deaths - počet úmrtí dětí do pěti let na 1000 obyvatel
* Polio - proočkovanost proti dětské obrně (Pol3) u dětí ve věku 1 roku (%)
* Total expenditure - Výdaje vládních institucí na zdravotnictví jako procento celkových vládních výdajů (%)
* Diphtheria - pokrytí očkováním proti záškrtu, tetanu a černému kašli (DTP3) u jednoletých dětí (%)
* HIV/AIDS - počet úmrtí na 1 000 živě narozených dětí na HIV/AIDS (0-4 roky)
* GDP - hrubý domácí produkt na obyvatele (v USD)
* Population - počet obyvatel země
* thinness 1-19 years - podíl dětí ve věku 10-19 let s indexem tělesné hmotnosti (BMI) menším než 2 směrodatné odchylky pod mediánem (%)
* thinness 5-9 years - podíl dětí ve věku 5-9 let s indexem tělesné hmotnosti (BMI) menším než 2 směrodatné odchylky pod mediánem (%)
* Income composition of resources - Index lidského rozvoje z hlediska příjmového složení zdrojů (index v rozmezí 0 až 1)
* Schooling - počet let školní docházky (roky)


## Pokyny k vypracování

**Body zadání**, za jejichž (poctivé) vypracování získáte **25 bodů**: 
  * V notebooku načtěte data ze souboru `data.csv`. Vhodným způsobem si je rozdělte na podmnožiny, které Vám poslouží pro trénování (trénovací), porovnávání modelů (validační) a následnou predikci výkonnosti finálního modelu (testovací).
    
  * Proveďte základní předzpracování dat:
    * Projděte si jednotlivé příznaky a transformujte je do vhodné podoby pro použití ve vybraném regresním modelu.
    * Nějakým způsobem se vypořádejte s chybějícími hodnotami. _Pozor na metodické chyby!_
    * Můžete využívat i vizualizace. Vše stručně ale náležitě komentujte.
<br /><br />
  * Vytvořte **vlastní implementaci náhodného lesa**. Použijte k tomu níže předpřipravenou kostru.
  
  * Na připravená data postupně aplikujte Vaši předchozí implementaci modelu náhodného lesa, dále jeden z modelů **lineární regrese** nebo **hřebenové regrese**, a alespoň jeden další model podle Vašeho uvážení, přičemž pro každý z těchto modelů přiměřeně:
    * Okomentujte vhodnost daného modelu pro daný typ úlohy.
    * Experimentujte s normalizací (standardizace/min-max), pokud pro daný model očekáváte její příznivý vliv.
    * Vyberte si hlavní hyperparametry k ladění a najděte jejich nejlepší hodnoty (vzhledem k RMSE).
    * Pro model s nejlepšími hodnotami hyperparametrů určete jeho chybu pomocí RMSE a MAE. _Pozor na metodické chyby!_
    * Získané výsledky vždy řádně okomentujte.
<br /><br />
  * Ze všech zkoušených možností v předchozím kroku vyberte finální model a odhadněte, jakou chybu (RMSE) můžete očekávat na nových datech, která jste doposud neměli k dispozici. _Pozor na metodické chyby!_
    
  * Nakonec načtěte vyhodnocovací data ze souboru `evaluation.csv`. Pomocí finálního modelu napočítejte predikce pro tato data. Vytvořte soubor `results.csv`, ve kterém získané predikce uložíte s využitím tří sloupců: **Country**, **Year** a **Life expectancy**. Tento soubor též odevzdejte (uložte do repozitáře vedle notebooku).

  * Ukázka prvních řádků souboru `results.csv`:
  
```
Country,Year,Life expectancy
Peru,2012,71.4
Peru,2013,72.6
...
```


## Poznámky k odevzdání

  * Řiďte se pokyny ze stránky https://courses.fit.cvut.cz/BI-ML1/homeworks/index.html.

In [1]:
import math
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
from random import randrange
from sklearn.tree import DecisionTreeRegressor
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import Ridge, LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error
from sklearn.neighbors import KNeighborsRegressor

import warnings

In [2]:
df = pd.read_csv("data.csv")

Náhodně rozdělím data na trénovací, testovací a validační.

In [3]:
Xdata = df.drop("Life expectancy", axis = 1)
Ydata = df["Life expectancy"]
rd_seed = 333
Xtrain, Xtest, Ytrain, Ytest = train_test_split(Xdata, Ydata, test_size=0.4, random_state=rd_seed) 
Xval, Xtest, Yval, Ytest = train_test_split(Xtest, Ytest, test_size=0.5, random_state=rd_seed) 


nektera data maji nesmyslne hodnoty - dalo by se je nějak upravit například zkoušet posouvat desetinnou čárku u dat, která jsou řádově mimo. Některá data nedávají smysl vůbec, například populace v afgánistánu.

In [4]:
pd.set_option('display.max_rows', None)
df.head(100)

,Country,Year,Status,Life expectancy,Adult Mortality,infant deaths,Alcohol,percentage expenditure,Hepatitis B,Measles,...,Polio,Total expenditure,Diphtheria,HIV/AIDS,GDP,Population,thinness 1-19 years,thinness 5-9 years,Income composition of resources,Schooling
0,Afghanistan,2015,Developing,65.0,263.0,62,0.01,71.279624,65.0,1154,...,6.0,8.16,65.0,0.1,584.259210,33736494.0,17.2,17.3,0.479,10.1
1,Afghanistan,2014,Developing,59.9,271.0,64,0.01,73.523582,62.0,492,...,58.0,8.18,62.0,0.1,612.696514,327582.0,17.5,17.5,0.476,10.0
2,Afghanistan,2013,Developing,59.9,268.0,66,0.01,73.219243,64.0,430,...,62.0,8.13,64.0,0.1,631.744976,31731688.0,17.7,17.7,0.470,9.9
3,Afghanistan,2012,Developing,59.5,272.0,69,0.01,78.184215,67.0,2787,...,67.0,8.52,67.0,0.1,669.959000,3696958.0,17.9,18.0,0.463,9.8
4,Afghanistan,2011,Developing,59.2,275.0,71,0.01,7.097109,68.0,3013,...,68.0,7.87,68.0,0.1,63.537231,2978599.0,18.2,18.2,0.454,9.5
5,Afghanistan,2010,Developing,58.8,279.0,74,0.01,79.679367,66.0,1989,...,66.0,9.20,66.0,0.1,553.328940,2883167.0,18.4,18.4,0.448,9.2
6,Afghanistan,2009,Developing,58.6,281.0,77,0.01,56.762217,63.0,2861,...,63.0,9.42,63.0,0.1,445.893298,284331.0,18.6,18.7,0.434,8.9
7,Afghanistan,2008,Developing,58.1,287.0,80,0.03,25.873925,64.0,1599,...,64.0,8.33,64.0,0.1,373.361116,2729431.0,18.8,18.9,0.433,8.7
8,Afghanistan,2007,Developing,57.5,295.0,82,0.02,10.910156,63.0,1141,...,63.0,6.73,63.0,0.1,369.835796,26616792.0,19.0,19.1,0.415,8.4
9,Afghanistan,2006,Developing,57.3,295.0,84,0.03,17.171518,64.0,1990,...,58.0,7.43,58.0,0.1,272.563770,2589345.0,19.2,19.3,0.405,8.1


### Nahrazení NaN  
dole seznam příznaků, kde se vyskytují nany a jak je potencionálně nahradit.

pokud má stát v příznaku víc než polovinu hodnot NaN, musím je zreplikovat od podobných států. Jinak vezmu medián ostatních hodot toho příznaku


Alcohol - south sudan je jediný, kde chybí dost - medián všech států

Hepatitis B - státy se stejnou dětskou úmrtností

Polio - státy se stejnou dětskou úmrtností

Total expenditure - státy se stejným developmentem

Diphtheria - státy se stejnou dětskou úmrtností

GDP - státy se stejným developmentem

Population - toto je podle mě nezávislé úplně - medián všech států

BMI - ve státech kde chybí víc jak půl, chybí i příznak thinness. Proto opužiju státy se stejným developmentem

thinness  1-19 years - státy se stejným developmentem

thinness 5-9 years - státy se stejným developmentem

Income composition of resources - median zemí se stejnou školní docházkou a developmentem

Schooling - median států se stejným percentage expenditure - když davájí peníze na zdravotnictvi tak snad i na školství



In [5]:
nan_cols = [i for i in df.columns if df[i].isnull().any()]
nan_cols


['Alcohol',
 'Hepatitis B',
 'BMI',
 'Polio',
 'Total expenditure',
 'Diphtheria',
 'GDP',
 'Population',
 'thinness  1-19 years',
 'thinness 5-9 years',
 'Income composition of resources',
 'Schooling']

In [6]:
# dostane dataset kde se jména států shodují - vrátí true pokud je v datech v příznaku column víc než půlka nan
def more_than_half_nulls(data, column):
    nulls = 0
    for i in range(data.shape[0]):
        if pd.isnull(data.iloc[i][column]):
            nulls += 1
    
    # bugfixing
    if nulls-2 > data.shape[0]/2:
        return True
    return False
    
    
def fill_nan_by_state_mean(mean_value, dataframe, state, column):
    mask = dataframe['Country'] == state

    dataframe.loc[mask, column] = dataframe.loc[mask, column].fillna(mean_value)
    return dataframe
  
    
# v datasetu nahradí všechny nan příznaky ve sloupci column u státu state
# k výpočtu používá trénovací data, což může trochu zhoršit čitelnost kódu
def substitue_nan(data, column, state):    
#     print(column, state)
    matching_rows = data[data["Country"] == state]
    
#     print("matching_rows = ", matching_rows.shape[0])
    if not matching_rows.empty:
        first_row = matching_rows.iloc[0]
    else:
        return data
    
    mean_value = None
    if more_than_half_nulls(Xtrain.loc[Xtrain["Country"] == state], column):
        if column == "Alcohol":
            print("here")
            mean_value = Xtrain[column].median()
        elif column == "Hepatitis B":
            mean_value = Xtrain.loc[Xtrain["infant deaths"] == first_row["infant deaths"]][column].median()
        elif column == "Polio":
            mean_value = Xtrain.loc[Xtrain["infant deaths"] == first_row["infant deaths"]][column].median()
        elif column == "BMI":
            mean_value = Xtrain.loc[Xtrain["Status"] == first_row["Status"]][column].median()
        elif column == "Total expenditure":
            mean_value = Xtrain.loc[Xtrain["Status"] == first_row["Status"]][column].median()
        elif column == "Diphtheria":
            mean_value = Xtrain.loc[Xtrain["infant deaths"] == first_row["infant deaths"]][column].median()
        elif column == "GDP":
            mean_value = Xtrain.loc[Xtrain["Status"] == first_row["Status"]][column].median()
        elif column == "Population":
            mean_value = Xtrain[column].median()
        elif column == "thinness  1-19 years":
            mean_value = Xtrain.loc[Xtrain["Status"] == first_row["Status"]][column].median()
        elif column == "thinness 5-9 years":
            mean_value = Xtrain.loc[Xtrain["Status"] == first_row["Status"]][column].median()
        elif column == "Income composition of resources":
            mean_value = Xtrain.loc[(Xtrain["Schooling"] == first_row["Schooling"]) & (Xtrain["Status"] == first_row["Status"])][column].median()
        elif column == "Schooling":
            mean_value = Xtrain.loc[Xtrain["percentage expenditure"] == first_row["percentage expenditure"]][column].median() 
        
    else:
        mean_value = Xtrain.loc[Xtrain["Country"] == state][column].median()
        

    data = fill_nan_by_state_mean(mean_value, data, state, column)
#     print("mean_value = ", mean_value)
    return data


def fill_mean(data):
    nan_cols = [i for i in data.columns if data[i].isnull().any()]
   
    for col in nan_cols:
        mean_v = Xtrain[col].median()
        data[col] = data[col].fillna(mean_v)
    return data


# dolní všechny nan
def substitue_all_nan(data):
    nan_cols = [i for i in data.columns if data[i].isnull().any()]
#     print("naser si", nan_cols)
    for col in nan_cols:
#         print(data[col])
        null_in_row = data[data[col].isnull()]
#         print(null_in_row)
        states = null_in_row["Country"].unique()
#         print(states)
        for state in states:
#             print(col, state)
            data = substitue_nan(data, col, state)

    # pro jistotu, aby nezustal nikde nan
    data = fill_mean(data)
    return data

# za všechny nany dosadí -1 = mohlo by být užitečné pro les, použiju jako hyperparametr
def fill_negative(data):
    data = data.fillna(-1)
    return data
        

### Číselné příznaky

In [7]:
pd.set_option('display.max_rows', None)
Xdata["Country"].unique()

array(['Afghanistan', 'Albania', 'Algeria', 'Angola',
       'Antigua and Barbuda', 'Argentina', 'Armenia', 'Australia',
       'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain', 'Bangladesh',
       'Barbados', 'Belarus', 'Belgium', 'Belize', 'Benin', 'Bhutan',
       'Bolivia (Plurinational State of)', 'Bosnia and Herzegovina',
       'Botswana', 'Brazil', 'Brunei Darussalam', 'Bulgaria',
       'Burkina Faso', 'Burundi', "Côte d'Ivoire", 'Cabo Verde',
       'Cambodia', 'Cameroon', 'Canada', 'Central African Republic',
       'Chad', 'Chile', 'China', 'Colombia', 'Comoros', 'Congo',
       'Costa Rica', 'Croatia', 'Cuba', 'Cyprus', 'Czechia',
       "Democratic People's Republic of Korea",
       'Democratic Republic of the Congo', 'Denmark', 'Djibouti',
       'Dominican Republic', 'Ecuador', 'Egypt', 'El Salvador',
       'Equatorial Guinea', 'Eritrea', 'Estonia', 'Ethiopia', 'Fiji',
       'Finland', 'France', 'Gabon', 'Gambia', 'Georgia', 'Germany',
       'Ghana', 'Greece', 'Grenad

Status má jenom dvě unikátní hodnoty, tak jsem se je rozhodl předělat na 0 a 1 bez použití dummy sloupce.

In [8]:
Xtrain["Status"].unique()

array(['Developing', 'Developed'], dtype=object)

In [9]:
def code_development(data):
    if data == "Developing":
         data = "0"
    elif data == "Developed":
        data = "1"
    return data

def convert_status(data):
    data["Status"] = data["Status"].apply(code_development)
    data["Status"] = data["Status"].astype(int)
    return data

In [10]:
asia = ['Afghanistan', 'Armenia', 'Azerbaijan', 'Bahrain', 'Bangladesh', 'Bhutan', 'Brunei Darussalam', 'Cambodia', 'China', 'Georgia', 'India', 'Indonesia', 'Iran (Islamic Republic of)', 'Iraq', 'Israel', 'Japan', 'Jordan', 'Kazakhstan', 'Kuwait', 'Kyrgyzstan', "Lao People's Democratic Republic", 'Lebanon', 'Malaysia', 'Maldives', 'Mongolia', 'Myanmar', 'Nepal', 'Oman', 'Pakistan', 'Qatar', 'Republic of Korea', 'Saudi Arabia', 'Singapore', 'Sri Lanka', 'Syrian Arab Republic', 'Tajikistan', 'Thailand', 'Timor-Leste', 'Turkey', 'Turkmenistan', 'United Arab Emirates', 'Uzbekistan', 'Viet Nam', 'Yemen']

europe = ['Albania', 'Austria', 'Belarus', 'Belgium', 'Bosnia and Herzegovina', 'Bulgaria', 'Croatia', 'Cyprus', 'Czechia', 'Denmark', 'Estonia', 'Finland', 'France', 'Georgia', 'Germany', 'Greece', 'Hungary', 'Iceland', 'Ireland', 'Italy', 'Latvia', 'Lithuania', 'Luxembourg', 'Malta', 'Republic of Moldova', 'Romania', 'Russian Federation', 'Serbia', 'Slovakia', 'Slovenia', 'Spain', 'Sweden', 'Switzerland', 'The former Yugoslav republic of Macedonia', 'United Kingdom of Great Britain and Northern Ireland']

africa = ['Algeria', 'Angola', 'Benin', 'Botswana', 'Burkina Faso', 'Burundi', "Côte d'Ivoire", 'Cabo Verde', 'Cameroon', 'Central African Republic', 'Chad', 'Comoros', 'Congo', 'Djibouti', 'Egypt', 'Equatorial Guinea', 'Eritrea', 'Eswatini', 'Ethiopia', 'Gabon', 'Gambia', 'Ghana', 'Guinea', 'Guinea-Bissau', 'Kenya', 'Lesotho', 'Liberia', 'Libya', 'Madagascar', 'Malawi', 'Mali', 'Mauritania', 'Mauritius', 'Morocco', 'Mozambique', 'Namibia', 'Niger', 'Nigeria', 'Rwanda', 'Sao Tome and Principe', 'Senegal', 'Seychelles', 'Sierra Leone', 'Somalia', 'South Africa', 'South Sudan', 'Sudan', 'Tanzania', 'Togo', 'Uganda', 'Zambia', 'Zimbabwe']

australia_oceania = ['Australia', 'Fiji', 'Kiribati', 'Marshall Islands', 'Micronesia (Federated States of)', 'Nauru', 'New Zealand', 'Palau', 'Papua New Guinea', 'Samoa', 'Solomon Islands', 'Tonga', 'Tuvalu', 'Vanuatu']

north_america = ['Antigua and Barbuda', 'Bahamas', 'Barbados', 'Belize', 'Canada', 'Costa Rica', 'Cuba', 'Dominican Republic', 'El Salvador', 'Grenada', 'Guatemala', 'Haiti', 'Honduras', 'Jamaica', 'Mexico', 'Nicaragua', 'Panama', 'Saint Lucia', 'Saint Vincent and the Grenadines', 'Trinidad and Tobago', 'United States of America']

south_america = ['Argentina', 'Bolivia (Plurinational State of)', 'Brazil', 'Chile', 'Colombia', 'Ecuador', 'Guyana', 'Paraguay', 'Peru', 'Suriname', 'Uruguay', 'Venezuela (Bolivarian Republic of)']


Pro velké množství unikátních hodnot příznaku country jsem se rozhodl je rozdělit do kategorií, pro které jsem vytvořil dummy příznaky. Poloha země na kontinentu by mohla hrát významnou roli pro věk dožití.

In [11]:
# přidá pro každý kontinent dummy příznak
def add_continent_dummies(data):
    return data.assign(asia=0, europe=0, africa=0, australia_oceania=0, north_america=0, south_america=0)
    
# rozdělí země do kontinentů (ty mam vytvořené výše v polích)  
def assign_country(data):
    for index, row_data in data.iterrows():
        if row_data["Country"] in asia:
            data.at[index, "asia"] = 1
        elif row_data["Country"] in europe:
            data.at[index, "europe"] = 1
        elif row_data["Country"] in africa:
            data.at[index, "africa"] = 1
        elif row_data["Country"] in australia_oceania:
            data.at[index, "australia_oceania"] = 1
        elif row_data["Country"] in north_america:
            data.at[index, "north_america"] = 1
        elif row_data["Country"] in south_america:
            data.at[index, "south_america"] = 1
    return data
        
def drop_country(data):
    data = data.drop('Country', axis=1)
    return data

def country_to_int(data):
    data = add_continent_dummies(data)
    data = assign_country(data)
    data = drop_country(data)
    return data
    

In [12]:
def transform_data(data):
    data = convert_status(data)
    data = country_to_int(data)
    return data

### Moje impementace random forest

In [13]:
# Váš kód zde


########################################################
# Předpřipravená kostra modelu náhodného lesa
class CustomRandomForest:
    trees = []

    # vytvořím prázdné stromy s jejich příslušnými parametry a uložím si je do pole
    def __init__(self, n_estimators, max_samples, max_depth, **kwargs):

        self.n_estimators = n_estimators
        self.max_samples = max_samples
        self.max_depth = max_depth
        self.kwargs = kwargs
        
        for i in range(self.n_estimators):
               self.trees.append(DecisionTreeRegressor(max_depth=self.max_depth, **kwargs)) 
        
    # vygeneruje n náhoných čísel mezi 0 až Xtrain velikost, kde n je velikost * max_samples parametr   
    def bootstrap(self, size) -> list:
        result = []
        values = size * self.max_samples
        for num in range(int(values)):
            result.append(randrange(size))
        
        return result
            
    
    # natrénuje každý strom
    def fit(self, X, y):
        if isinstance(X, pd.DataFrame):
            X = X.values
 
        for tree in self.trees:
            positions = self.bootstrap(X.shape[0])
            
            xtmp = X[positions]
            ytmp = y.iloc[positions]
            
            tree.fit(xtmp, ytmp)
            


    def predict_val(self, X) -> int:
        result = []

        if isinstance(X, pd.DataFrame):
            X = X.values  # Convert DataFrame to NumPy array

        # Ensure that the shape is (1, -1) for a single sample with any number of features
        X = X.reshape(1, -1)

        for tree in self.trees:
            prediction = tree.predict(X)
            result.append(prediction)

        res = np.mean(result)

        return res
    
        
    def predict(self, X):
        if isinstance(X, pd.DataFrame):
            X = X.values
        
        # vytvořím pole, do kterého uložím výsledek
        y_predicted = np.zeros((X.shape[0],))
        
        for num in range(X.shape[0]):
            tmp = self.predict_val(X[num, :])

            y_predicted[num] = tmp
        
        return y_predicted


### Trénování modelů

Vytvořím si 2 sady dat s různým řešením nan, které budu používat na modely.

In [14]:
Xtrain_m = substitue_all_nan(Xtrain)
Xval_m = substitue_all_nan(Xval)
Xtest_m = substitue_all_nan(Xtest)

Xtrain_n = fill_negative(Xtrain)
Xval_n = fill_negative(Xval)
Xtest_n = fill_negative(Xtest)

here


Převedu příznaky na číselné

In [15]:
Xtrain_m = transform_data(Xtrain_m)
Xval_m = transform_data(Xval_m)
Xtest_m = transform_data(Xtest_m)

Xtrain_n = transform_data(Xtrain_n)
Xval_n = transform_data(Xval_n)
Xtest_n = transform_data(Xtest_n)

Vyberu si jeden z datasetů, který budu používat a zkusím ho normalizovat i standardizovat

In [16]:
scaler_mm = MinMaxScaler()
scaler_std = StandardScaler()

# current_data_train = Xtrain_m
# current_data_val = Xval_m
# current_data_test = Xtest_m
current_data_train = Xtrain_n
current_data_val = Xval_n
current_data_test = Xtest_n

scaler_mm.fit(current_data_train)
scaler_std.fit(current_data_train)

X_train_minmax = scaler_mm.transform(current_data_train)
X_train_std = scaler_std.transform(current_data_train)
X_val_minmax = scaler_mm.transform(current_data_val)
X_val_std = scaler_std.transform(current_data_val)
X_test_minmax = scaler_mm.transform(current_data_test)
X_test_std = scaler_std.transform(current_data_test)


In [36]:
Yval.shape
X_val_minmax.shape

(544, 26)

### Hřebenová regrese 
lineární jsem usoudil že nebude efektivní, kvůli závislosti příznaků

Vyzkouším různé hodnoty hyperparametru lambda a interceptu. Dále jako hyperparametr použiju, zda budu pracovat se škálováním dat

In [37]:
lam = 1
intercept = False
best_rmse = 10000000
# (0-no scaler, 1-minmax, 2-standard)
best_data = 0

all_data = [current_data_train, X_train_minmax, X_train_std]
all_val = [current_data_val, X_val_minmax, X_val_std]

for i in range(0, 35):
    inter = True
    for j in range(2):
        for dat in range(3):
            clf_ridge = Ridge(alpha=i, fit_intercept=inter)
            clf_ridge.fit(all_data[dat], Ytrain)
            RMSE_val_ridge = mean_squared_error(Yval, clf_ridge.predict(all_val[dat]), squared = False)
            
            if RMSE_val_ridge < best_rmse:
                best_rmse = RMSE_val_ridge
                lam = i
                intercept = inter
                best_data = dat
            inter = False


/home/mara/.local/lib/python3.11/site-packages/sklearn/linear_model/_ridge.py:200: LinAlgWarning: Ill-conditioned matrix (rcond=3.15968e-18): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T
/home/mara/.local/lib/python3.11/site-packages/sklearn/linear_model/_ridge.py:200: LinAlgWarning: Ill-conditioned matrix (rcond=3.03861e-18): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T
/home/mara/.local/lib/python3.11/site-packages/sklearn/linear_model/_ridge.py:200: LinAlgWarning: Ill-conditioned matrix (rcond=3.38045e-18): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T
/home/mara/.local/lib/python3.11/site-packages/sklearn/linear_model/_ridge.py:200: LinAlgWarning: Ill-conditioned matrix (rcond=3.2512e-18): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T
/home/mara/.local/lib/python3.11/site-packages/sklearn/linear

/home/mara/.local/lib/python3.11/site-packages/sklearn/linear_model/_ridge.py:200: LinAlgWarning: Ill-conditioned matrix (rcond=8.17312e-18): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T
/home/mara/.local/lib/python3.11/site-packages/sklearn/linear_model/_ridge.py:200: LinAlgWarning: Ill-conditioned matrix (rcond=8.70747e-18): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T
/home/mara/.local/lib/python3.11/site-packages/sklearn/linear_model/_ridge.py:200: LinAlgWarning: Ill-conditioned matrix (rcond=8.38889e-18): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T
/home/mara/.local/lib/python3.11/site-packages/sklearn/linear_model/_ridge.py:200: LinAlgWarning: Ill-conditioned matrix (rcond=8.93113e-18): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T
/home/mara/.local/lib/python3.11/site-packages/sklearn/linea

In [39]:
print(best_rmse, lam, intercept, best_data)


3.8465523637160643 5 False 0


výstup při nahrazení nanu -1

nejlepší výstup: 3.837246865761791 lambda = 6 intercept = False neupravená data (0)

vypočítání nanu

nejlepší výstup: 3.8465523637160643 lambda = 5 intercept = False neupravená data (0)


### Náhodný les

Níže jsou dva pokusy o natrénování lesa s ruznými hyperparametry. V prvním jsem se zajímal jen o počet stromů a z části i o hloubku. V druhém jsem zkoumal efekt i v závislosti na velikosti množiny na které stromy trénuju a škálování vstupních dat.

Kvůli značné časové náročnosti už jsem nezkoumal další hyperparametry jako například max_features

In [21]:
estimators = 0
data_portion = 0
best_depth = 0

best_rmse = 10000000

# (0-no scaler, 1-minmax, 2-standard)
best_data = 0
dat = 1

all_data = [Xtrain, X_train_minmax, X_train_std]
all_val = [Xval, X_val_minmax, X_val_std]

for est in range(155, 180, 15):
    for depth in range(20, 24, 2):
        forest = CustomRandomForest(n_estimators=est, max_samples=0.3, max_depth=depth)
        forest.fit(all_data[dat], Ytrain)
        RMSE_val_ridge = mean_squared_error(Yval, forest.predict(all_val[dat]), squared = False)

        if RMSE_val_ridge < best_rmse:
            best_rmse = RMSE_val_ridge
            estimators = est
            data_portion = 0.3
            best_depth = depth
            best_data = dat
        
    print(est)

155
170


In [22]:
print(best_rmse, estimators, data_portion, best_depth, best_data)

2.3318129196492023 170 0.3 20 1


výstup při nahrazení nanu -1 
2.327442610581602 145 0.3 22 1

nejlepší výstup: 2.327442610581602 počet stromů = 145 podíl dat = 0.3 hloubka = 22 normalizovaná data (1)

nany vypočítány funkcí

nejlepší výstup: 2.3472484907423614 počet stromů = 215 podíl dat = 0.3 hloubka = 18 normalizovaná data (1)

Zde zkoumám jak výsledek ovlivňují i ostatní hyperparametry, které volím s hodnotami kolem nejlepších hodnot z předchozího měření

In [23]:
estimators = 0
data_portion = 0
best_depth = 0

best_rmse = 10000000

# (0-no scaler, 1-minmax, 2-standard)
best_data = 0

all_data = [current_data_train, X_train_minmax, X_train_std]
all_val = [current_data_val , X_val_minmax, X_val_std]

for est in range(135, 150, 5):
    for depth in range(17, 21):
        for samples in np.arange(0.4, 0.6, 0.2):
            for dat in range(3):
                forest = CustomRandomForest(n_estimators=est, max_samples=samples, max_depth=depth)
                forest.fit(all_data[dat], Ytrain)
                RMSE_val_ridge = mean_squared_error(Yval, forest.predict(all_val[dat]), squared = False)
                
                if RMSE_val_ridge < best_rmse:
                    best_rmse = RMSE_val_ridge
                    estimators = est
                    data_portion = samples
                    best_depth = depth
                    best_data = dat
    print(est)

135
140
145


In [24]:
print(best_rmse, estimators, data_portion, best_depth, best_data)

2.212086059209918 135 0.4 18 2


nahrazení nanu -1

nejlepší výstup: 2.212086059209918 počet stromů = 135 podíl dat = 0.4 hloubka = 18 standardizovaná data (2)

nany vypočítány funkcí

nejlepší výstup: 2.2248201161385905 počet stromů = 215 podíl dat = 0.4 hloubka = 20 neupravená data (0)

### KNN

Jako poslední zkusim knn model kde budu zkoumat vliv hyperparametrů počet sousedů, metrika (jen eukleidovská geometrie), vliv vzdálenosti sousedů od zkoumaného bodu a škálování.

In [24]:
# https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsRegressor.html
# data std, minmax a default

best_rmse = 1000000
# (0-no scaler, 1-minmax, 2-standard)
best_data = 5
best_neighbours = 0
best_weights = None # uniform / distance
best_metrics = 1

all_data = [current_data_train, X_train_minmax, X_train_std]
all_val = [current_data_val, X_val_minmax, X_val_std]


for neigh in range(1, 50):
    for met in range(1, 3): # might try more
        for weight in ["uniform", "distance"]:
            knn = KNeighborsRegressor(n_neighbors=neigh, p=met, weights=weight)
            for i in range(3):
                knn.fit(all_data[i], Ytrain)
                RMSE_val_ridge = mean_squared_error(Yval, knn.predict(all_val[i]), squared = False)

                if RMSE_val_ridge < best_rmse:
                    best_data = i
                    best_neighbours = neigh
                    best_weights = weight
                    best_metrics = met
                    best_rmse = RMSE_val_ridge
            


In [25]:
print(best_rmse, best_neighbours, best_data, best_metrics, best_weights)

2.4140499569915668 4 2 1 distance


nany nahrazeny -1

nejlepší výstup: 2.4426835268754292 sousedi= 3 standardizovaná data (2) metrika = 1 váha = distance

nany vypočítány funkcí

nejlepší výstup: 2.4140499569915668 sousedi = 4 standardizovaná data (2) metrika = 1 váha = distance

nejlépe skóruje les na datech, kde nany jsou nahrazené -1 a standardizované

In [25]:
forest = CustomRandomForest(n_estimators=135, max_samples=0.4, max_depth=18)
forest.fit(X_train_std, Ytrain)


In [26]:
print("rmse = ", mean_squared_error(Ytest, forest.predict(X_test_std), squared = False))
print("mae = ", mean_absolute_error(Ytest, forest.predict(X_test_std)))

rmse =  2.056789927095609
mae =  1.2985884988941516


### Finální předpověď

In [36]:
evaluation = pd.read_csv("evaluation.csv")
result = evaluation[["Country", "Year"]]

Použiju na data úpravu jako na původní data, pro vzor mediánů použiju trénovací data

In [37]:
evaluation = fill_negative(evaluation)
evaluation = transform_data(evaluation)
evaluation = scaler_std.transform(evaluation)


In [38]:
life_expect = forest.predict(evaluation) 

In [39]:
result.insert(2, "Life expectancy", 0)
result["Life expectancy"] = life_expect

In [40]:
result.to_csv("results.csv", index=False)

In [41]:
df = pd.read_csv("results.csv")

In [42]:
df.head(10)

,Country,Year,Life expectancy
0,Azerbaijan,2015,73.908737
1,Azerbaijan,2014,74.588797
2,Azerbaijan,2013,74.644639
3,Azerbaijan,2012,74.225243
4,Azerbaijan,2011,74.304221
5,Azerbaijan,2010,74.143968
6,Belgium,2015,83.682758
7,Belgium,2014,83.344365
8,Belgium,2013,82.680875
9,Belgium,2012,82.784583
